### PR TITLE
kernel: Add k_reschedule()

### LINKED ISSUE #79361: EDF updates don't work when made from ISR context
**Describe the bug**

The EDF scheduler is not working as expected when the deadlines are updated within ISRs. The deadline values get correctly updated, but the threads don't execute in the expected order. I have first noticed something was going on on #79124, but after a bit of digging through the kernel files I'm convinced it's an actual bug.

**To Reproduce**

Run this code on your desired environment, preferably one with 64-bit timers (e.g. RISC-V targets). It might look lengthy, but it's actually quite simple. What it does:

- creates two identical threads, each with a private message queue;
- creates a master timer which periodically sends a message to each thread (.e.g.: every two seconds). The threads are unblocked when they receive a message, execute their cycle and get back to waiting for new messages to repeat it all.

Basically the timer ensures the threads are periodic, activating and sending the messages at every specified period. On every activation a new absolute deadline is generated for each thread. All extra context can be found within the comments.

Kconfig options:
```c
CONFIG_SYS_CLOCK_TICKS_PER_SEC=10000
CONFIG_SCHED_DEADLINE=y
CONFIG_LOG=y       /* makes printk be handled asynchronously by a background thread (reducing the overhead) */

# QEMU only
CONFIG_QEMU_ICOUNT=n
CONFIG_STDOUT_CONSOLE=y
```

source code:
```c
#include <zephyr/kernel.h>
/*
	Remember that for "pure" EDF results, all
	user threads must have the same static priority.
	That happens because EDF will be a tie-breaker
	among two or more ready tasks of the same static
	priority. An arbitrary positive number is chosen here.
*/

#define EDF_PRIORITY 5
#define INACTIVE -1
#define MSEC_TO_CYC(msec)       ((msec * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) / MSEC_PER_SEC)
#define MSEC_TO_USEC(msec)      (msec * 1000)

typedef struct {
    int id;
    int32_t rel_deadline_msec;
    int32_t period_msec;
    uint32_t wcet_msec;
    k_tid_t thread;
    struct k_msgq queue;
} edf_t;

// edf-related structs with the properties for each thread ////////////////////////////////

edf_t task1 = {
	.id = 1,
	.rel_deadline_msec = 10,
	.wcet_msec = 1
};

edf_t task2 = {
	.id = 2,
	.rel_deadline_msec = 50,
	.wcet_msec = 5
};

// timer-related functions ///////////////////////////////////////////////////////////////////

void grow(edf_t *task){
	task->wcet_msec += 1;
	task->rel_deadline_msec += 10;
}


void trigger(edf_t *task){
	const char t = 1;
	k_msgq_put(&task->queue, &t, K_NO_WAIT);
	int deadline = (int) MSEC_TO_CYC(task->rel_deadline_msec);
	k_thread_deadline_set(task->thread, deadline);
}


void trigger_all(struct k_timer *timer){
	printk("\n");
	trigger(&task1);
	trigger(&task2);
	grow(&task1);   //task 1 progressively gains +10ms of relative deadline
}

// thread function /////////////////////////////////////////////////////////////////////////

void thread_function(void *task_props, void *a2, void *a3){
	char buffer[10];
	char message;
	uint32_t deadline = 0;
	uint64_t start = 0;
	uint64_t end = 0;

	edf_t *task = (edf_t *) task_props;
	task->thread = k_current_get();
	k_msgq_init(&task->queue, buffer, sizeof(char), 10);
	printk("[ %d ] ready.\n", task->id);

	for(;;){
		/*
			The periodic thread has no period, ironically.
			What makes it periodic is the master timer
			associated with it that regularly sends new
			messages to the thread queue (e.g. every two
			seconds).
		*/
		k_msgq_get(&task->queue, &message, K_FOREVER);
		start = k_cycle_get_64();
		deadline = task->thread->base.prio_deadline;
		k_busy_wait(MSEC_TO_USEC(task->wcet_msec));
		end = k_cycle_get_64();
		printk("[ %d ]\tstart %llu\t end %llu\t dead %u\n", task->id, start, end, deadline);
	}
}

// defines and main //////////////////////////////////////////////////////////////////////

K_THREAD_DEFINE(
	task1_thread, 2048,
	thread_function,
	&task1, NULL, NULL,
	EDF_PRIORITY, 0, INACTIVE
);


K_THREAD_DEFINE(
	task2_thread, 2048,
	thread_function,
	&task2, NULL, NULL,
	EDF_PRIORITY, 0, INACTIVE
);


K_TIMER_DEFINE(master_timer, trigger_all, NULL);

int main(void){
         k_sleep(K_SECONDS(1));
        printk("\nexecuting on: %s\n\n", CONFIG_BOARD_TARGET);

	k_thread_start(task1_thread);
	k_thread_start(task2_thread);
	k_sleep(K_SECONDS(1));
	
	k_timer_start(&master_timer, K_SECONDS(1), K_SECONDS(2));
	return 0;
}
```

Congratulations, now your EDF should be NOT working! which is why I started this bug report in the first place.

**Expected behavior**
Thread 1 starts with a relative deadline of 10ms and thread 2 of 50ms, so when both are fired at the same time thread 1 should take the lead and execute first. However, since thread 1 is subject to the `grow()` function inside `trigger_all()`, it will gain +10ms of relative deadline at every trigger. So the expected outcome is that thread 2 eventually takes the lead:
```c
[ 1 ]   start 30149527   end 30171369    dead 30237806       // <-- thread 1 has the earliest deadline
[ 2 ]   start 30175266   end 30225378    dead 30645602

// a few cycles later...

[ 2 ]   start 130116686  end 130166893   dead 130613698
[ 1 ]   start 130167576  end 130237728   dead 130713353    // <-- no longer the earliest deadline
```

**Observed behavior**
Thread 2 remains the first one to run at ALL times, regardless of the deadline values. That happens, apparently, because thread 2 is the last inside ``trigger_all()` to be, well, triggered - so the scheduler basically behaves like a LIFO. If we trigger thread 2 and then 1, the execution order would be 1-2 forever.
```c
[ 2 ]   start 30112071   end 30162828    dead 30611068
[ 1 ]   start 30164161   end 30184363    dead 30209562     // <- but this is the earliest deadline... :(

[ 2 ]   start 50121647   end 50171696    dead 50621379
[ 1 ]   start 50171874   end 50201914    dead 50321308    // <- here aswell
```

**Impact**
This bug prevents periodic EDF threads from being timely updated, so it limits the usage of this algorithm in the application. For me, specifically, it is a showstopper for the feature #78000 that I've opened an RFC a few weeks ago, since it depends on updating its own thread's deadline inside a timer callback to work accordingly. 

**Environment (please complete the following information):**

 - OS: Windows 11 23H2
 - Zephyr 3.7.99, SDK 0.16.8
 - Commit SHA: main is at caf86128b75

### PR DESCRIPTION
**Pull request by peter-mitsis**
Adds the routine k_reschedule() to force a schedule point. It is callable at both thread and ISR level.

Most applications will never use this.

Fixes #79361

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -1050,6 +1050,25 @@ static inline void z_vrfy_k_thread_deadline_set(k_tid_t tid, int deadline)
...
 #endif /* CONFIG_SCHED_DEADLINE */
 
+void z_impl_k_reschedule(void)
+{
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&_sched_spinlock);
+
+	update_cache(0);
+	flag_ipi(IPI_ALL_CPUS_MASK);
```
Comment (andyross) [inline kernel/sched.c:1060]: Broadcasting an IPI here seems very heavyweight.  The use case at hand is just getting the scheduler to do a reschedule on ISR exit, nothing to do with SMP behavior.  Basically this implements "do an immediate reschedule on all cores in the system", which is a lot more work than required.

File: include/zephyr/kernel.h [hunk 2]
```diff
@@ -946,6 +946,23 @@ __syscall void k_thread_priority_set(k_tid_t thread, int prio);
...
+ *
+ * This routine invokes the scheduler to force a schedule point. If invoked
+ * from within a thread, the scheduler will be invoked immediately (provided
+ * interrupts were not locked when invoked). If invoked from within an ISR,
+ * the scheduler will be invoked upon exiting the ISR.
+ *
+ * Invoking the scheduler allows the kernel to make an immediate determination
+ * as to what the next thread to execute should be. This is not the same as
+ * yielding for that voluntarily surrenders control of the CPU to another
+ * thread of equal or greater priority.
```
Comment (andyross) [inline include/zephyr/kernel.h:960]: Might need more clarity here.  The distinction the reader wants to understand is that yield will always choose a different thread of the same priority to run, and this won't.

**Review by andyross — COMMENTED** [review]
Nitpicking on that IPI.  I get why you're including it, it's a "reschedule", which implies that we need to be sure scheduler state is reflected on all CPUs.  But it's really expensive, especially for things like the actual bug being fixed where the app just has a thread it wants to adjust an EDF priority on.  And now it's synchronously interrupting everything!

My gut says that (1) k_yield is already sufficient for the thread use case and we don't need this there and (2) we'd be better served with a lightweight "force ISR to reschedule on exit" primitive for the specific issue.

Apps that really truly genuinely want to scramble the universe for their event already have access to `arch_sched_broadcast_ipi()`.

**Comment by alexpaschoaletto** [comment]
> Nitpicking on that IPI. I get why you're including it, it's a "reschedule", which implies that we need to be sure scheduler state is reflected on all CPUs. But it's really expensive, especially for things like the actual bug being fixed where the app just has a thread it wants to adjust an EDF priority on. And now it's synchronously interrupting everything!

I agree. I'm no SMP expert so my suggestion might be nonsense, but what if the rescheduling could be restricted to the core that invoked it?

Anyway, thank you @peter-mitsis for the initiative!

**Comment by peter-mitsis** [comment]
Great points! Changes made.

**Review by andyross — APPROVED** [review]
Looks great here.

**Review by cfriedt — APPROVED** [review]

**Pull request merged**